File: engine/src/main/java/com/mjaseem/trick/strategy/HighCardStrategy.java
package com.mjaseem.trick.strategy;

import com.mjaseem.trick.engine.Card;
import com.mjaseem.trick.engine.GameHistory;
import com.mjaseem.trick.engine.Records;
import com.mjaseem.trick.engine.Trick;

import java.util.Comparator;
import java.util.List;
import java.util.stream.IntStream;

// High-card strategy implementation
public class HighCardStrategy implements Strategy {

    @Override
    public int chooseCard(List<Card> hand, GameHistory history, Trick trick, Records.Suit trumpSuit) {
        return IntStream.range(0, hand.size()).boxed()
                .filter(i -> trick.plays().isEmpty() || hand.get(i).suit().equals(trick.plays().getFirst().getValue().suit()))
                .max(Comparator.comparing(i -> hand.get(i).rank()))
                .or(() -> IntStream.range(0, hand.size()).boxed().min(Comparator.comparing(i -> hand.get(i).rank())))
                .orElseThrow();

    }

}

File: rl/strategy/high_card_strategy.py
from typing import List

from engine import Strategy, Card, Trick
from engine.card import Suit


class HighCardStrategy(Strategy):
    def choose_card(
            self, hand: List[Card], game_history: List['Trick'], trick: Trick, trump_suit: Suit
    ) -> int:
        """
        Select the highest-ranked card of the leading suit if possible; otherwise, play the lowest-ranked card.
        """
        # Check if there is a leading suit in the current trick
        leading_suit = trick.plays[0][1].suit if trick.plays else None

        # Try to find the highest-ranked card in the leading suit
        high_card_index = (
            max(
                (i for i in range(len(hand)) if not leading_suit or hand[i].suit == leading_suit),
                key=lambda i: hand[i].rank,
                default=None,
            )
        )

        # If no card of the leading suit is found, play the lowest-ranked card
        if high_card_index is not None:
            return high_card_index

        # Fallback to the lowest-ranked card in the hand
        return min(range(len(hand)), key=lambda i: hand[i].rank)
